### ISSUE TITLE
Kokkos Promotion 2.6.00 > 2.7.00

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#1620]
kokkos-promotion branch of Trilinos created to accumulate fixes for promotion. 
Created from SHA ea50776573f163acc58d63b6184cef51612f355d
Author: Matthias Mayr <mmayr@sandia.gov>
Date:   Wed Apr 18 12:46:09 2018 -0700

One commit pushed to that branch with ifpack2 fix related to the StaticCrsGraph changes to allow for memory traits template parameter.

**Comment by ndellingwood** [comment]
Linking to trilinos/Trilinos#2735 and adding link to wiki page of deprecated code to remove:
https://github.com/kokkos/kokkos/wiki/DeprecationPage

**Comment by ndellingwood** [comment]
**White Cuda** results - deprecated code enabled
- updated kokkos-promotion branch to fix typos in a few sacado and stokhos tests. 

Trilinos pristine: develop branch
SHA 75b13d9d1e975ab7c25ca175e07cde1d47536bd3

Trilinos updated: kokkos-promotion branch
SHA (prior to snapshots) 3e117039ca40e2056380c2d33de0d6594e83cb57

kokkos: develop branch
SHA 04790a6

kokkos-kernels: develop branch
SHA 45481c358a973d6de467b3307b6e3beee009ba3a

Build errors:
Same build errors on both branches:
ROL
` error: the type in a dynamic_cast must be a pointer or reference to a complete class type, or void *`
TrilinosCouplings (file not found in MueLu) 
`fatal error: MueLu_AztecEpetraOperator.hpp: No such file or directory`

Test failures:
2 new tests failed with kokkos-promotion branch while 1 test passed that failed with the pristine branch.
updated:
```
MueLu_UnitTestsTpetra_kokkos_MPI_1 (Failed)
MueLu_UnitTestsTpetra_kokkos_MPI_4 (Failed)
```

pristine (failure not in updated):
`Intrepid_test_Discretization_Basis_HGRAD_TET_Cn_FEM_Test_02_MPI_1 (Failed)`

Tests not run:
Same on both branches
`ROL_test_function_BlockOperatorCheck_MPI_1 (Not Run)`

**Comment by ndellingwood** [comment]
Still need to look at the 2 failing MueLu tests with Cuda. 
The many Pthreads tests not run with the updated build are due to the old enable macro used in the Trilinos integration testing configure scripts:
`Kokkos_ENABLE_Pthread=${PTHREAD}`

wIth the updated build this should be
`Kokkos_ENABLE_Threads=${PTHREAD}`

Things are actually in pretty good shape testing with deprecated code enabled :)

@crtrott You mentioned starting on test_all_sandia type scripts for the Trilinos integration testing, would you like me to wait for those or should I update the config scripts in Trilinos with option to disable deprecated code and update with the `Kokkos_ENABLE_Threads` macro? Updates to the kokkos and kokkos-kernels submission scripts will also be needed to take input argument to disable deprecated code.

**Comment by ndellingwood** [comment]
MueLu test with Cuda that failed:
`Aggregates_kokkos_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_JustDist2UncoupledAggregation_UnitTest`


Debugger output indicates an issue in this function call
`BuildAggregatesDistance2`

After running mangled debugger output through c++filt:
`MueLu::AggregationPhase1Algorithm_kokkos<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::BuildAggregatesDistance2(MueLu::LWGraph_kokkos<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> > const&, MueLu::Aggregates_kokkos<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >&, std::vector<unsigned int, std::allocator<unsigned int> >&, int&, int) const`


cuda-gdb output shows an illegal address must have been accessed
```
1. Aggregates_kokkos_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_JustDist2UncoupledAggregation_UnitTest ...
CUDA Exception: Device Illegal Address
The exception was triggered in device 0.

Program received signal CUDA_EXCEPTION_10, Device Illegal Address.
[Switching focus to CUDA kernel 0, grid 36, block (0,0,0), thread (0,0,0), device 0, sm 12, warp 0, lane 0]
```


Runtime output with stack trace:
```
bash-4.2$ mpirun -n 1 ./MueLu_UnitTests_kokkos.exe
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name white23 and rank 0!
Kokkos::Cuda::initialize WARNING: running kernels compiled for compute capability 3.5 on device with compute capability 3.7 , this will likely reduce potential performance.

***
*** Unit test suite ...
***


Sorting tests by group name then by the order they were added ... (time = 0.000111)

Running unit tests ...

0. Aggregates_kokkos_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_JustUncoupledAggregation_UnitTest ... Using default factory (MueLu::CoalesceDropFactory) for building 'Graph'.
Using default factory (MueLu::AmalgamationFactory) for building 'UnAmalgamationInfo'.
Build (MueLu::UncoupledAggregationFactory_kokkos)
 Build (MueLu::CoalesceDropFactory_kokkos)
  Build (MueLu::AmalgamationFactory)

   ******* WARNING *******
   AmalagamationFactory::Build(): no striding information available. Use blockdim=1 with offset=0

  ******* WARNING *******
  lightweight wrap is deprecated
  algorithm = "classical": threshold = 0, blocksize = 1
  Detected 0 Dirichlet nodes

  ******* WARNING *******
  Level::Set: unable to store "DofsPerNode" generated by factory 0x10036f1c2c0 on level 0, as it has not been requested and no keep flags were set for it

  ******* WARNING *******
  Level::Set: unable to store "A" generated by factory 0x10036f1c2c0 on level 0, as it has not been requested and no keep flags were set for it
 Build (MueLu::CoalesceDropFactory)
  Build (MueLu::AmalgamationFactory)

   ******* WARNING *******
   AmalagamationFactory::Build(): no striding information available. Use blockdim=1 with offset=0
  lightweight wrap = 1
  algorithm = "classical": threshold = 0, blocksize = 1

  ******* WARNING *******
  Level::Set: unable to store "Filtering" generated by factory 0x1004a0b4220 on level 0, as it has not been requested and no keep flags were set for it
  Detected 0 Dirichlet nodes

  ******* WARNING *******
  Level::Set: unable to store "Graph" generated by factory 0x1004a0b4220 on level 0, as it has not been requested and no keep flags were set for it
  Number of dropped entries in unamalgamated matrix graph: 0/43 (0%)
 Algo "Phase - (Dirichlet)"
  BuildAggregates (Phase - (Dirichlet))
    aggregated : 0 (phase), 0/15 [0.00%] (total)
    remaining  : 15
    aggregates : 0 (phase), 0 (total)
 Algo "Phase 1 (main)"
  BuildAggregates (Phase 1 (main))
    aggregated : 15 (phase), 15/15 [100.00%] (total)
    remaining  : 0
    aggregates : 5 (phase), 5 (total)
 Algo "Phase 2a (secondary)"
  BuildAggregates (Phase 2a (secondary))
    aggregated : 0 (phase), 15/15 [100.00%] (total)
    remaining  : 0
    aggregates : 0 (phase), 5 (total)
 Algo "Phase 2b (expansion)"
  BuildAggregates (Phase 2b (expansion))
    aggregated : 0 (phase), 15/15 [100.00%] (total)
    remaining  : 0
    aggregates : 0 (phase), 5 (total)
 Algo "Phase 3 (cleanup)"
  BuildAggregates (Phase 3 (cleanup))
    aggregated : 0 (phase), 15/15 [100.00%] (total)
    remaining  : 0
    aggregates : 0 (phase), 5 (total)
 "UC": MueLu::Aggregates_kokkos{nGlobalAggregates = 5}
[Passed] (0.0206 sec)
1. Aggregates_kokkos_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_JustDist2UncoupledAggregation_UnitTest ... terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaDeviceSynchronize() error( cudaErrorIllegalAddress): an illegal memory access was encountered /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/trilinos-update/packages/kokkos/core/src/Cuda/Kokkos_Cuda_Impl.cpp:119
Traceback functionality not available

[white23:43734] *** Process received signal ***
[white23:43734] Signal: Aborted (6)
[white23:43734] Signal code:  (-6)
[white23:43734] [ 0] [0x100000050478]
[white23:43734] [ 1] /lib64/libc.so.6(abort+0x280)[0x100020780d70]
[white23:43734] [ 2] /home/projects/pwr8-rhel73-lsf/gcc/5.4.0/lib64/libstdc++.so.6(_ZN9__gnu_cxx27__verbose_terminate_handlerEv+0x1f4)[0x1000204cb174]
[white23:43734] [ 3] /home/projects/pwr8-rhel73-lsf/gcc/5.4.0/lib64/libstdc++.so.6(+0xa7b44)[0x1000204c7b44]
[white23:43734] [ 4] /home/projects/pwr8-rhel73-lsf/gcc/5.4.0/lib64/libstdc++.so.6(+0xa61c8)[0x1000204c61c8]
[white23:43734] [ 5] /home/projects/pwr8-rhel73-lsf/gcc/5.4.0/lib64/libstdc++.so.6(__gxx_personality_v0+0x2b8)[0x1000204c7078]
[white23:43734] [ 6] /home/projects/pwr8-rhel73-lsf/gcc/5.4.0/lib64/libgcc_s.so.1(+0xbc54)[0x10002070bc54]
[white23:43734] [ 7] /home/projects/pwr8-rhel73-lsf/gcc/5.4.0/lib64/libgcc_s.so.1(_Unwind_Resume+0x174)[0x10002070c844]
[white23:43734] [ 8] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/kokkos/core/src/libkokkoscore.so.12(_ZN6Kokkos4Impl25cuda_internal_error_throwE9cudaErrorPKcS3_i+0x610)[0x10001ffaf460]
[white23:43734] [ 9] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/kokkos/core/src/libkokkoscore.so.12(_ZN6Kokkos4Impl23cuda_device_synchronizeEv+0x3c)[0x10001ffaf51c]
[white23:43734] [10] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/kokkos/core/src/libkokkoscore.so.12(_ZN6Kokkos4Cuda5fenceEv+0x18)[0x10001ffaf558]
[white23:43734] [11] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/kokkos/core/src/libkokkoscore.so.12(_ZNK6Kokkos12CudaUVMSpace10deallocateEPvm+0x20)[0x10001ffab080]
[white23:43734] [12] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/kokkos/core/src/libkokkoscore.so.12(_ZN6Kokkos4Impl22SharedAllocationRecordINS_12CudaUVMSpaceEvED2Ev+0x5c)[0x10001ffab19c]
[white23:43734] [13] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/muelu/adapters/libmuelu-adapters.so.12(_ZN6Kokkos4Impl22SharedAllocationRecordINS_12CudaUVMSpaceENS0_16ViewValueFunctorINS_4CudaEiLb1EEEED0Ev+0x2c)[0x10000b81022c]
[white23:43734] [14] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/muelu/src/libmuelu.so.12(_ZN6Kokkos4Impl98_GLOBAL__N__74_tmpxft_00000f27_00000000_7_MueLu_AggregationPhase1Algorithm_kokkos_cpp1_ii_77b64a9010deallocateINS_12CudaUVMSpaceENS0_16ViewValueFunctorINS_4CudaEiLb1EEEEEvPNS0_22SharedAllocationRecordIvvEE+0x34)[0x10000f3cacf4]
[white23:43734] [15] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/kokkos/core/src/libkokkoscore.so.12(_ZN6Kokkos4Impl22SharedAllocationRecordIvvE9decrementEPS2_+0x550)[0x10001ff9e2b0]
[white23:43734] [16] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/muelu/src/libmuelu.so.12(_ZNK5MueLu33AggregationPhase1Algorithm_kokkosIiiN6Kokkos6Compat23KokkosDeviceWrapperNodeINS1_4CudaENS1_12CudaUVMSpaceEEEE24BuildAggregatesDistance2ERKNS_14LWGraph_kokkosIiiS6_EERNS_17Aggregates_kokkosIiiS6_EERSt6vectorIjSaIjEERii+0xf48)[0x10000f3eff58]
[white23:43734] [17] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/muelu/src/libmuelu.so.12(_ZNK5MueLu33AggregationPhase1Algorithm_kokkosIiiN6Kokkos6Compat23KokkosDeviceWrapperNodeINS1_4CudaENS1_12CudaUVMSpaceEEEE15BuildAggregatesERKN7Teuchos13ParameterListERKNS_14LWGraph_kokkosIiiS6_EERNS_17Aggregates_kokkosIiiS6_EERSt6vectorIjSaIjEERi+0x340)[0x10000f3f0440]
[white23:43734] [18] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/muelu/src/libmuelu.so.12(_ZNK5MueLu34UncoupledAggregationFactory_kokkosIiiN6Kokkos6Compat23KokkosDeviceWrapperNodeINS1_4CudaENS1_12CudaUVMSpaceEEEE5BuildERNS_5LevelE+0x164c)[0x10000f4fe44c]
[white23:43734] [19] ./MueLu_UnitTests_kokkos.exe(_ZNK10MueLuTests56Aggregates_kokkos_JustDist2UncoupledAggregation_UnitTestIdiiN6Kokkos6Compat23KokkosDeviceWrapperNodeINS1_4CudaENS1_12CudaUVMSpaceEEEE15runUnitTestImplERN7Teuchos18basic_FancyOStreamIcSt11char_traitsIcEEERb+0xcf8)[0x1039b478]
[white23:43734] [20] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/teuchos/core/src/libteuchoscore.so.12(_ZNK7Teuchos12UnitTestBase11runUnitTestERNS_18basic_FancyOStreamIcSt11char_traitsIcEEE+0x5c)[0x10001ff00e9c]
[white23:43734] [21] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/teuchos/core/src/libteuchoscore.so.12(_ZN7Teuchos18UnitTestRepository15runUnitTestImplERKNS_12UnitTestBaseERNS_18basic_FancyOStreamIcSt11char_traitsIcEEE+0x6c)[0x10001ff0434c]
[white23:43734] [22] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/teuchos/core/src/libteuchoscore.so.12(_ZN7Teuchos18UnitTestRepository12runUnitTestsERNS_18basic_FancyOStreamIcSt11char_traitsIcEEE+0x126c)[0x10001ff06a6c]
[white23:43734] [23] /ascldap/users/ndellin/IntegrationTests-White/cuda-fix/build_updated/packages/teuchos/core/src/libteuchoscore.so.12(_ZN7Teuchos18UnitTestRepository20runUnitTestsFromMainEiPPc+0x138)[0x10001ff089a8]
[white23:43734] [24] ./MueLu_UnitTests_kokkos.exe(main+0x80)[0x101d0cc0]
[white23:43734] [25] /lib64/libc.so.6(+0x24700)[0x100020764700]
[white23:43734] [26] /lib64/libc.so.6(__libc_start_main+0xc4)[0x1000207648f4]
[white23:43734] *** End of error message ***
```

**Comment by ibaned** [comment]
@crtrott maybe we should remove TrilinosCouplings from our promotion testing? The Trilinos folks don't seem to keep it clean.

**Comment by mhoemmen** [comment]
@ibaned I think Sierra cares about TrilinosCouplings.  On the other hand, what TrilinosCouplings actually builds and tests depends radically on build settings -- the package has a huge number of optional dependencies.

**Comment by ndellingwood** [comment]
@ibaned I'll try that. The Cuda failures with the kokkos-promotion branch for the tests
```
MueLu_UnitTestsTpetra_kokkos_MPI_1 (Failed)
MueLu_UnitTestsTpetra_kokkos_MPI_4 (Failed)
```
are a bit trickier, chasing those down now and not sure if it's in MueLu or KokkosKernels.

**Comment by ndellingwood** [comment]
I have a fix for MueLu for the tests:
```
MueLu_UnitTestsTpetra_kokkos_MPI_1 (Failed)
MueLu_UnitTestsTpetra_kokkos_MPI_4 (Failed)
```
@ibaned the kokkos-tools kernel-logger is pretty awesome for tracking down problems, thanks for suggestion!

I only built the MueLu unit tests, will rebuild to test the macro-fix for TrilinosCouplings then push the fixes.

**Comment by ndellingwood** [comment]
Updated comment on trilinos/Trilinos#2786 regarding the `MueLu_AztecEpetraOperator.hpp: No such file or directory` error; @lucbv pointed out guards are better placed in the `CMakeLists.txt` file based on the packages that are required for this to compile (ifdef-ing out the `#include MueLu_AztecEpetraOperator.hpp` in the file may cause other problems, since MueLu and Epetra and AztecOO are each enabled). 
Since this is broken on both pristine and updated builds, @cgcgcg can work on this separately from the integration testing and we can merge it in.

**Comment by ndellingwood** [comment]
Cuda testing has 100% MueLu tests passing with the changes; Threads testing has 100% Tpetra tests passing with the changes (many tests not build previously). Using these as 'spot-checks' I think the next round of integration testing is ready to begin, will merge (no-ff) the Trilinos develop branch into kokkos-promotion and restart the tests, hoping all the issues are resolved.

**Comment by ndellingwood** [comment]
Round 2 integration test results look good - fewer build errors and same/fewer test failures for the updated branch  :)


**Shepard serial**
* updated branch makes it farther through the build, same test error on both branches. 
updated branch
build errors:
One Muelu error
`catastrophic error: cannot open source file "MueLu_AztecEpetraOperator.hpp"`
This will be resolved by PR trilinos/Trilinos#2794 - thanks @lucbv and @cgcgc !

test failures:
100% tests ran
`ML_MLP_Restart_MPI_4 (Failed)`


pristine branch
build errors:
Several errors like below, caused build to terminate early.
```icpc: error #10106: Fatal error in /home/projects/x86-64/intel/compilers/2017/compilers_and_libraries_2017.4.196/linux/bin/intel64/mcpcom, terminated by kill signal
compilation aborted for /ascldap/users/ndellin/IntegrationTests-Shepard/serial/trilinos-pristine/packages/muelu/example/advanced/stratimikos/Stratimikos.cpp (code 1)
make[2]: *** [packages/muelu/example/advanced/stratimikos/CMakeFiles/MueLu_Stratimikos.dir/build.make:62: packages/muelu/example/advanced/stratimikos/CMakeFiles/MueLu_Stratimikos.dir/Stratimikos.cpp.o] Error 1```

test failures:
146 tests not ran (due to early compilation termination)
`ML_MLP_Restart_MPI_4 (Failed)`


**Shepard threads**
updated branch
build errors:
Same MueLu error discussed in serial build.

test failures:
No test failures :)

tests not run: 5
```
Xpetra_BlockedCrsMatrix_UnitTests_MPI_4 (Not Run)
Xpetra_CrsMatrix_UnitTests_MPI_4 (Not Run)
Xpetra_MatrixMatrix_UnitTests_MPI_4 (Not Run)
Xpetra_Matrix_UnitTests_MPI_4 (Not Run)
Xpetra_ThyraBlockedOperator_UnitTests_MPI_4 (Not Run)
```

pristine branch
build errors:
`compilation error: Out of memory asking for 67112960`, occurs 92% through build.

test failures:
`ML_MLP_Restart_MPI_4 (Failed)`

tests not run: 143
This includes the Xpetra tests not run on the updated branch.


**White openmp**
updated branch
build errors:
Same MueLu error discussed in serial build.

test failures:
67 failures, same as on pristine branch, all tests run.

pristine branch
build errors:
Same MueLu error discussed in serial build.

test failures:
67 failures, all tests run.


**White cuda**
updated branch
build errors:
Same MueLu error discussed in serial build.
ROL - `packages/rol/test/function/test_03.cpp(79): error: the type in a dynamic_cast must be a pointer or reference to a complete class type, or void *`

test failures:
68 failures, same as on pristine branch

test not run:
`ROL_test_function_BlockOperatorCheck_MPI_1 (Not Run)`

pristine branch
build errors:
Same MueLu error discussed in serial build.
ROL - `packages/rol/test/function/test_03.cpp(79): error: the type in a dynamic_cast must be a pointer or reference to a complete class type, or void *`

test failures:
68 failures, all tests run.

test not run:
`ROL_test_function_BlockOperatorCheck_MPI_1 (Not Run)`

**Comment by ndellingwood** [comment]
@crtrott @ibaned testing with deprecated code enabled is complete.

**Comment by ndellingwood** [comment]
Integration testing in good shape; PR #1638 a recently discovered blocker and resolution of #1631.

**Comment by ndellingwood** [comment]
PR #1638 merged, issue #1631 requires better understanding and does not impact this promotion.

**Comment by ndellingwood** [comment]
PR trilinos/Trilinos#2822 submitted for with Kokkos + KokkosKernels promotion.
@crtrott @ibaned @srajama1 @kyungjoo-kim @mhoemmen

**Comment by ndellingwood** [comment]
PR trilinos/Trilinos#2822 merged, kokkos and kokkos-kernels promotion 2.7.00 complete - thanks all for your help and contributing!

**Comment by mhoemmen** [comment]
Thanks @ndellingwood ! :D Great work!!! :D
